### ISSUE TITLE
V4 Tilling Sprites

### ISSUE THREAD

**Issue by Greewi** [pixijs/pixi.js#2660]
Hello,

We are experiencing a new issue with the tilling sprites in PIXI V4.

Here is a fiddle that show the issue : https://jsfiddle.net/sa0zj0vq/1/

When the scale is set to 0.5 (line 6 of the fiddle), holes appear in the repeating pattern. If the scale is set to 1, the issue disappear. We looked into the code of the V4 and we didn't find any difference with our V3 version which works well in this case.

**Comment by ivanpopelyshev** [comment]
May be its in shader:

TilingSprite trick with using "%" operation inside shader is VERY bad, because it wont work for linear filtering, so its better to get rid of it when its possible (when texture rect is covering all the basetexture)

Had v3 version the same trick ?

**Comment by Greewi** [comment]
Yes, it has the same trick. The point is that we use a texture atlas so our tilling texture is only a part of a bigger texture. As you said we had problem with linear filtering, that's why the texture coord is clamped with a 1 pixel margin (uPixelSize) in the next shader instruction.

**Comment by Greewi** [comment]
Ok... I Found it. (Following an intuition)

It is caused by the generation of the mipmaps. If I set PIXI.MIPMAP_TEXTURES to false the bug disappear. I don't see a way to fix it with elegance...

**Comment by ivanpopelyshev** [comment]
@Greewi that too. "%" is that kind of operator :(

**Comment by ivanpopelyshev** [comment]
So, we can do the following:

1. If mipmaps are on, and texture covers all basetexture, just use other kind of shader
2. If mipmaps are on and texture is not covering basetexture, make a warning in the console, ask to disable mipmapping for this particular baseTexture.

**Comment by Greewi** [comment]
I have not found how to disable the mipmapping for a single texture. For compressed texture we can do it when generating our DDS/PVR/ATC (for compressed textures the mipmaps are not computed on the fly but generated in the file itself). Unfortunately, many devices doesn't handle compressed textures (for instance, apple phones require square textures (and the results may sometime look ugly as PVR is not designed for "cartoonish" images).

**Comment by JiDW** [comment]
In addition, it means that you can't use spritesheet for your tiling sprite images, as it would remove the mipmaps for the whole texture, not only the frame you need to repeat.

**Comment by ivanpopelyshev** [comment]
@JiDW Exactly! Btw, I'm so glad you are here :+1: how is your new project?

**Comment by ivanpopelyshev** [comment]
@Greewi easy. 

```js
resources['spritesheet'].textures['someFrame.png'].baseTexture.mipmap = false;
```

**Comment by Greewi** [comment]
Nice. But it has to be done before having rendered the texture. By the way, we still have the problem descriped by JiDW.

**Comment by JiDW** [comment]
@ivanpopelyshev Thanks, it is going well, if not for this bug. Greewi and I work together ;)

**Comment by ivanpopelyshev** [comment]
@JiDW I have the same but in RPGMaker MV, so I will be glad to fix it too.

**Comment by ivanpopelyshev** [comment]
@JiDW Can you give me specs of devices that your old game runs perfectly smooth (60fps)? I want to buy a number of devices for testing

**Comment by englercj** [comment]
Can you retest this with the latest dev build and see if that has fixed your issue?

You can see how to download the latest build [in the FAQs](https://github.com/pixijs/pixi.js/wiki/FAQs#where-can-i-get-a-build).

Thanks!

**Comment by Greewi** [comment]
Hello, I missed your previous message.

Unfortunately, the problem has not been fixed. Here is a fiddle using PIXI 4.3 : https://jsfiddle.net/sa0zj0vq/6/

**Comment by Greewi** [comment]
Let's get a functionnal fiddle (the previous has a https error) : https://jsfiddle.net/sa0zj0vq/9/

**Comment by Greewi** [comment]
We did a few tests on the fiddle : no matter how we change the scale, the bug still persist. (That was not the case in the 4.0.0 version).
A fiddle using 4.0.0 and scale 1 : https://jsfiddle.net/sa0zj0vq/11/
A fiddle using 4.3.0 and scale 1 : https://jsfiddle.net/sa0zj0vq/12/

Both version have a probleme with scale at 0.5.

**Comment by JiDW** [comment]
So it's worse now :D 

Anyone knows why ?

**Comment by ivanpopelyshev** [comment]
@JiDW because there's an option for that, and the clamping must be different for different textures. That's for your case:

```
  tilingSprite.uvTransform.clampMargin = 0;
  tilingSprite.uvTransform.clampOffset = -1;
```

That's if you want it to apply to ALL tilingsprites:

```
texture.transform = new PIXI.extras.TextureTransform(texture, 0);
texture.transform.clampMargin = 0;
```

However, current version affects X coord that way too, so beware. I think its better if we separate X clamp and Y clamp..

**Comment by JiDW** [comment]
Hello @ivanpopelyshev !

Thanks for your help, your explanation helped us to get the same result we had before 4.1.

However, the first issue was still there.

We managed to fix it, even while keeping the mipmaps by using an advanced feature in TexturePacker => Extrude

http://www.serenity-ff.com/images/screenshots/screenshot1484144038.png

http://www.serenity-ff.com/images/screenshots/screenshot1484144118.png

May it help other people with the same issue.

**Comment by GoodBoyDigital** [comment]
Hi there! Closing this issue for now due to the solution provided. Feel free to give us a poke if you would like this issue reopened. Thanks 👍

**Comment by DanielVyskoc** [comment]
This is still not solved for me. Using texturepacker extrude is not convenient solution.

**Comment by ivanpopelyshev** [comment]
There are multiple solutions. It depends on your case, and that problem is "solved" in pixi in a meaning that its possible to configure TilingSprite to do anything. If you dont understand how it works and docs are not enough.. well.. post your case in separate issue and we'll help.

**Comment by DanielVyskoc** [comment]
Hi, I did not read first post properly, but I think my problem is caused by same bug. I have similar problem with **normal** sprites. I am placing individual sprites next to each other (like tiles). There appear lines between sprites. This only happens when I use PIXI.Texture.formFrame() and sprite is scaled.

**Comment by Greewi** [comment]
Theses bugs a caused by interpolation : if you use a sprite directly adjacent to an other, there will always be a color bleeding as the sub pixels rendered are interpolated.

The mains solutions are the same :  use a single sprite textures or add a margin that extend the color of the border to each sprite. You can also try to use a nearest interpolation instead of a linear, but that will have an effect of the appearance of upscaled textures.

**Comment by ivanpopelyshev** [comment]
@Greewi is right about normal sprites. I can add one more alternative: you can use  https://github.com/pixijs/pixi-picture , though that renderer doesnt have batching at all. 

If your images are big and there are only few of them (like special effects in rpgmaker) then extra drawcalls wont hurt your app.
